docs/graph_engine/data_spec.rst

```diff
@@ -12,14 +12,93 @@ DeepGNN supports both homogeneous and heterogeneous graphs. Nodes and Edges supp
 Graph Data Format
 *****************
 
-DeepGNN supports two file formats: JSON and TSV.
+DeepGNN supports three file formats: EdgeList, JSON and TSV.
 Users can generate a graph in either format then our pipeline will convert it into binary for training and inference.
 DeepGNN also supports writing custom decoders, see [the decoders file](https://github.com/microsoft/DeepGNN/blob/main/src/python/deepgnn/graph_engine/snark/decoders.py).
 Just inheret the base class Decoder, overwrite the decode function and pass the new decoder as an argument to the converter or dispatcher.
 
-1. `JSON <#json-format>`_: Heterogeneous or homegeneous graph.
+1. `EdgeList <#EdgeList-format>`_: Heterogeneous or homegeneous graph.
 
-2. `TSV <#tsv-format>`_: Homogeneous graph only.
+2. `JSON <#json-format>`_: Heterogeneous or homegeneous graph.
+
+3. `TSV <#tsv-format>`_: Homogeneous graph only.
+
+EdgeList Format
+===============
+
+The EdgeList format,
+* Supports heterogeneous and homegeneous graphs.
+* Nodes and edges are on separate lines, therefore a node and its outgoing edges may be on separate lines.
+* Small files that are fast to create and convert.
+
+`graph.csv` layout,
+
+.. code-block:: text
+	<node info>
+	<edge_1_info>
+	<edge_2_info>
+	...
+
+node_info: node_id,-1,node_type,node_weight,<features>
+edge_info: src,dst,edge_type,edge_weight,<features>
+features[dense]: dtype_name,length,v1,v2,...,dtype_name2,length2,v1,v2,...
+features[sparse with 2 dim coordinates vector]: dtype_name,values.size/coords.shape[1],c1,c2,...,v1,v2,...
+features[sparse with 1 dim coordinates vector]: dtype_name,values.size/0,c1,c2,...,v1,v2,...
+* Sort file as node, then all of the nodes outgoing edges sorted by type then dst.
```

Comment:
Lexographic sorting is problematic for the edge ordering. Edges must be sorted first by type then dst -- with dst being sorted in same order as nodes. Users can use ETL jobs to sort this file format which makes it much easier.

Reply:
> Lexographic sorting is problematic for the edge ordering.

Please, explain, why is problematic, why can't we change columns and make everything more natural?

> Users can use ETL jobs to sort this file format which makes it much easier.

We don't need any ETL pipelines, just a command line utility `sort` to work with lexicographic ordering: it implements an external sort algorithm and can process almost any file.

Reply:
Blocker is edge sort, edges need to be sorted by dst in increasing order of internal id, so if the user keeps track of this ordering the linear format will work for that scenario.

Reply:
Why would this make it work?

Reply:
Because you can sort edges/nodes in a lexicographic order before starting conversion.

Reply:
Why does reordering the columns enable lexicographic sorting?

Reply:
Current implementation:
```
# src, dst, type, data
1,3,2,
1,-1,1,
1,2,2,
```
Is hard to sort in the proposed format. Changed order before sorting:
```
# src, type, dst, data
1,2,3,
1,1,-1,
1,2,2,
```
after:
```
# src, type, dst, data
1,1,-1
1,2,2,
1,2,3,
```

Reply:
This works: sort edgelist.csv -t, -n -k1,1 -k3,3 -k2,2 --parallel=1
- You have to specify the column order to sort anyways for a csv so we don't need to actually switch them. I added to docs. But this does require you to invert the type of the node.
-

Reply:
Lexicographical order is way faster:
```sh
time sort ~/tmp/edges.txt -t, -n > ~/tmp/lexi.txt

real    1m46.391s
user    6m53.227s
sys     0m7.406s
```
vs
```sh
time sort ~/tmp/edges.txt -t, -n -k1,1 -k3,3 -k2,2 > ~/tmp/custom.txt

real    2m46.870s
user    9m47.953s
sys     0m8.420s
```

Reply:
lexographical sort does not work for non trivial files:
1,3,0,
1,-1,-1,
1,3,11,
1,2,11,
1,1,11,
1,0,11,
1,2,2,
2,-1,-5,
2,1,1

Reply:
Set env variable `LC_COLLATE=C` or `LC_ALL=C` to sort negative numbers properly.

Reply:
I'm saying if you run lexographical sort on that file, it will not work, it acts like each line is just one big number

Reply:
Hmm, ok. So why can't we use a natural column order: `1, 2, 3` and have to use `1, 3, 2`?

Reply:
src and dst are the only columns that are strictly required, the rest can be defaulted and skipped. So it makes sense that src and dst goes first and everything else after.

Reply:
We are talking about positions, not field names.
* If users have a default edge type, then sorting is based on columns `1, 2` (src and dst).
* If there are more than one type, then types should go before dst and sorting should be `1, 2, 3`, not `1, 3, 2`.

Reply:
Every sort function makes it easy to choose which columns have priority. It is confusing to everyone when the column order has to change like that, instead we keep it extensible
src, dst
src, dst, type, weight, features

Reply:
> Every sort function makes it easy to choose which columns have priority. 

It is easy to forget that edges should be sorted by type first. We already made [such mistake](https://github.com/microsoft/DeepGNN/commit/dd9c7c32ff9209d829004ad17242e37929fa70da) and format should help us to avoid it.

> It is confusing to everyone when the column order has to change like that, instead we keep it extensible. 

Not at all, the actual format is:
`src, type[optional], dst, weight[optional], features[optional]`.

Reply:
We can add an assertion to prevent confusion

Reply:
1. We are talking about data format, not code. You can't add assertion to data.
2. This thread is growing way to long and so far you didn't show any benefits of changing a natural column order. `src`/`type`/`destination` is kind of how edges are stored internally, so it will be always easier to keep input in a similar format.

Reply:
1. We already have an assertion in linear decoder to warn the user if they do this, we need the assertion either way to ensure their data is sorted.
2. This format makes sense
```
src, dst, type[optional], weight[optional], features[optional]
```
It does not make sense to have an optional variable in the middle of required variables, this is confusing
```
src, type[optional], dst, weight[optional], features[optional]
```
3. We already have the decoder in src,dst,type format, it will be easier to keep this input in the same format.
4. Edge  binaries are written in src,dst,type format, it will be easier to keep this input in the same format: https://github.com/microsoft/DeepGNN/blob/cdieckhaus/externalize_linear/src/python/deepgnn/graph_engine/snark/converter/writers.py#:~:text=%22%3DQQif%22%2C,weight%2C

Reply:
1. I am talking about data format, not code. Why are you talking about assertions?
2. It seems you are using python style for passing parameters. Optional arguments are totaly fine anywhere in the message. See [protocol-buffers for example](https://developers.google.com/protocol-buffers/docs/overview#syntax)
3. Decoder is actually broken, code is easier to fix than data. Here is an example. If we want to do a quick conversion with just a binary writer then we need to use something like this with `[src, dst, type, weight, feature]` order:
```python
from deepgnn.graph_engine.snark.converter.writers import BinaryWriter
bw = BinaryWriter(".", 0, True, True)
edges = []
with open("graph.txt", "r") as f:
    lines = f.readlines()
    raw_edges = [[int(x) for x  in line.split(",")] for line in lines]
    edges = list(map(lambda x: (x[0], x[2], x[1], 1, []), raw_edges))
dst_sorted = sorted(edges, key=lambda x: x[1])
tp_sorted = sorted(dst_sorted, key=lambda x: x[2])
src_sorted = sorted(tp_sorted, key=lambda x: x[0])
bw.add(src_sorted)
bw.close()
```
However, if we move type to a second place `[src, type, dst, weight, feature]` code will be much shorter:
```python
from deepgnn.graph_engine.snark.converter.writers import BinaryWriter
bw = BinaryWriter(".", 0, True, True)
edges = []
with open("graph.txt", "r") as f:
    lines = f.readlines()
    raw_edges = [[int(x) for x  in line.split(",")] for line in lines]
    edges = list(map(lambda x: (x[0], x[2], x[1], 1, []), raw_edges))
bw.add(sorted(edges))
bw.close()
```
4. If you read the comments to that code, you'll see that we were trying to avoid using extra memory. Graph engine actually transforms that data to [CSR format](https://github.com/microsoft/DeepGNN/blob/main/src/cc/lib/graph/partition.h#L118) where we store sorted types first and then all destinations. I don't know yet how to convert directly to CSR format on the fly yet and that's why we have temporary files with both dst and tp.

Reply:
#2 I'm saying it's confusing to explain to people, as opposed to saying just add an extra comma you have to tell them to put it between the two. For the sort in #3, its actually sorted(edges, key=lambda x: (x[0], x[2], x[1])).
But I actually tried to implement this and its not possible // easy. We need to know if dst == -1 or not, if the given line is a node or not, in order to determine whether or not there is a default value for the type given // whether dst is in column 2 or 3.

Reply:
why can't we use type=-1 to encode a node item?

Reply:
I'll do this, it will also mean the sort wont require you to invert node types